pkg/binary/envoy/controlplane/istio_test.go

```diff
@@ -38,39 +37,57 @@ func TestMain(m *testing.M) {
 	envoytest.FetchEnvoyAndRun(m)
 }
 
+// TestConnectsToMockPilotAsAGateway ensures Istio bootstraps Envoy configuration, and that configuration in
+// "networking.istio.io/v1alpha3" rules end up pushed back to Envoy via XDS.
+//
+// Technically what happens is bootstrap results in a "xds-grpc" cluster pointed at Pilot's gRPC address. During
+// bootstrap, Envoy invokes "/envoy.service.discovery.v3.AggregatedDiscoveryService/StreamAggregatedResources", which
+// reads configuration in pilot (from testdata/configs.yaml). This test passes when that configuration merges with the
+// bootstrap configuration generated by Istio.
 func TestConnectsToMockPilotAsAGateway(t *testing.T) {
-	pilotGrpcAddr, stopPilot := requireMockPilot(t, "testdata")
+	// only includes configs.yaml, which includes egress rules for googleapis
+	pilotServer, stopPilot := requireMockPilot(t, "testdata")
 	defer stopPilot()
 
+	pilotGrpc := pilotServer.GRPCListener.Addr().String()
 	runtime, err := envoy.NewRuntime(
 		func(r *envoy.Runtime) {
 			r.Config.Mode = envoy.ParseMode("loadbalancer")
-			r.Config.XDSAddress = pilotGrpcAddr
-			r.Config.IPAddresses = []string{"1.1.1.1"}
+			// Choosing a random port doesn't help because envoy_bootstrap.json hard-codes 15020 15021 and 15090
+			r.Config.AdminPort = 15000
+			r.Config.XDSAddress = pilotGrpc
+			r.Config.IPAddresses = []string{"127.0.0.1"} // prevent calling controlplane.retrieveIPs() on CI hosts
 		},
 		Istio,                                // this integrates and ends up reading pilotGrpcAddr
 		debug.EnableEnvoyAdminDataCollection, // this allows us to read clusters.txt
 	)
 	require.NoError(t, err, "error creating envoy runtime")
 	defer os.RemoveAll(runtime.DebugStore())
+	time.Sleep(5 * time.Second)
```

Comment:
I'm worried that this could be another source of flakiness.. but yeah I don't think of better alternative..

Reply:
uhhh nothing to see here.. move along.. 😊